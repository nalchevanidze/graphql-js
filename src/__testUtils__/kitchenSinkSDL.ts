export const kitchenSinkSDL = `
"""This is a description of the schema as a whole."""
schema {
  query: QueryType
  mutation: MutationType
  subscription: SubscriptionType
}

"""
This is a description
of the \`Foo\` type.
"""
type Foo implements Bar & Baz & Two {
  "Description of the \`one\` field."
  one: Type
  """This is a description of the \`two\` field."""
  two(
    """This is a description of the \`argument\` argument."""
    argument: InputType!
  ): Type
  """This is a description of the \`three\` field."""
  three(argument: InputType, other: String): Int
  four(argument: String = "string"): String
  five(argument: [String] = ["string", "string"]): String
  six(argument: InputType = {key: "value"}): Type
  seven(argument: Int = null): Type
}

type AnnotatedObject @onObject(arg: "value") {
  annotatedField(arg: Type = "default" @onArgumentDefinition): Type @onField
}

type UndefinedType

interface Bar {
  one: Type
  four(argument: String = "string"): String
}

interface AnnotatedInterface @onInterface {
  annotatedField(arg: Type @onArgumentDefinition): Type @onField
}

interface UndefinedInterface

interface Baz implements Bar & Two {
  one: Type
  two(argument: InputType!): Type
  four(argument: String = "string"): String
}

resolver Feed = Story | Article | Advert | Photo | Video

resolver AnnotatedUnion @onUnion = A | B

resolver AnnotatedUnionTwo @onUnion = A | B

resolver UndefinedUnion

scalar CustomScalar

scalar AnnotatedScalar @onScalar

enum Site {
  """This is a description of the \`DESKTOP\` value"""
  DESKTOP
  """This is a description of the \`MOBILE\` value"""
  MOBILE
  "This is a description of the \`WEB\` value"
  WEB
  VR
}

enum AnnotatedEnum @onEnum {
  ANNOTATED_VALUE @onEnumValue
  OTHER_VALUE
}

enum UndefinedEnum

data InputType = {
  key: String!
  answer: Int = 42
  other: Float = 1.23e4 @onInputFieldDefinition
}

data AnnotatedInput @onInputObject = {
  annotatedField: Type @onInputFieldDefinition
}

data UndefinedInput

directive @include(if: Boolean!) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

directive @myRepeatableDir(name: String!) repeatable on OBJECT | INTERFACE

`;
