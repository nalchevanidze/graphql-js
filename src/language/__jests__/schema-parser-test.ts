import { dedent } from '../../__testUtils__/dedent';
import {
  expectJSON,
  expectToThrowJSON,
  toJSONDeep,
} from '../../__testUtils__/expectJSON';

import { parse } from '../parser';

function expectSyntaxError(text: string) {
  return expectToThrowJSON(() => parse(text));
}

const JSONSnapshot = (doc: unknown) => expect(toJSONDeep(doc)).toMatchSnapshot();

describe('Schema Parser', () => {
  it('Simple type', () => {
    const doc = parse(dedent`
      resolver Hello {
        world: String
      }
    `);
    JSONSnapshot(doc);
  });

  it('parses type with description string', () => {
    const doc = parse(dedent`
      "Description"
      type Hello {
        world: String
      }
    `);
    JSONSnapshot(doc)
  });

  it('parses type with description multi-line string', () => {
    const doc = parse(dedent`
      """
      Description
      """
      # Even with comments between them
      type Hello {
        world: String
      }
    `);

    JSONSnapshot(doc)
  });

  it('Description followed by something other than type system definition throws', () => {
    expectSyntaxError('"Description" 1').to.deep.equal({
      locations: [{ column: 15, line: 1 }],
      message: 'Syntax Error: Unexpected Int "1".',
    });
  });

  it('Simple non-null type', () => {
    const doc = parse(dedent`
      type Hello {
        world: String!
      }
    `);

    JSONSnapshot(doc)
  });

  it('Single value Enum', () => {
    const doc = parse('data Hello = WORLD');

    JSONSnapshot(doc)
  });

  it('Double value Enum', () => {
    const doc = parse('data Hello = WO | RLD');

    JSONSnapshot(doc);
  });

  it('Simple field with arg', () => {
    const doc = parse(dedent`
      type Hello {
        world(flag: Boolean): String
      }
    `);

    JSONSnapshot(doc)
  });

  it('Simple field with arg with default value', () => {
    const doc = parse(dedent`
      type Hello {
        world(flag: Boolean = true): String
      }
    `);

    JSONSnapshot(doc);
  });

  it('Simple field with list arg', () => {
    const doc = parse(dedent`
      type Hello {
        world(things: [String]): String
      }
    `);

    JSONSnapshot(doc);
  });

  it('Simple field with two args', () => {
    const doc = parse(dedent`
      type Hello {
        world(argOne: Boolean, argTwo: Int): String
      }
    `);
    JSONSnapshot(doc);
  });

  it('Simple resolver', () => {
    const doc = parse('resolver Hello = World');
    JSONSnapshot(doc);
  });

  it('Union with two types', () => {
    const doc = parse('resolver Hello = Wo | Rld');
    JSONSnapshot(doc);
  });

  it('Union with two types and leading pipe', () => {
    const doc = parse('resolver Hello = | Wo | Rld');
    JSONSnapshot(doc);
  });

  it('Union fails with no types', () => {
    expectSyntaxError('resolver Hello = |').to.deep.equal({
      message: 'Syntax Error: Expected Name, found <EOF>.',
      locations: [{ line: 1, column: 19 }],
    });
  });

  it('Union fails with leading double pipe', () => {
    expectSyntaxError('resolver Hello = || Wo | Rld').to.deep.equal({
      message: 'Syntax Error: Expected Name, found "|".',
      locations: [{ line: 1, column: 19 }],
    });
  });

  it('Union fails with double pipe', () => {
    expectSyntaxError('resolver Hello = Wo || Rld').to.deep.equal({
      message: 'Syntax Error: Expected Name, found "|".',
      locations: [{ line: 1, column: 22 }],
    });
  });

  it('Union fails with trailing pipe', () => {
    expectSyntaxError('resolver Hello = | Wo | Rld |').to.deep.equal({
      message: 'Syntax Error: Expected Name, found <EOF>.',
      locations: [{ line: 1, column: 30 }],
    });
  });

  it('Scalar', () => {
    const doc = parse('scalar Hello');

    expectJSON(doc);
  });

  it('Simple data object', () => {
    const doc = parse(`
    data Hello {
      world: String
    }`);

    JSONSnapshot(doc)
  });

  it('Simple data object with args should fail', () => {
    expectSyntaxError(`
      data  Hello {
        world(foo: Int): String
      }
    `).to.deep.equal({
      message: 'Syntax Error: Expected ":", found "(".',
      locations: [{ line: 3, column: 14 }],
    });
  });
});
