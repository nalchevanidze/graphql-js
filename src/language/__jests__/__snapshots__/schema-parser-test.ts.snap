// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema Parser Simple type 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "ResolverTypeDefinition",
      "loc": Object {
        "end": 34,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 14,
          "start": 9,
        },
        "value": "Hello",
      },
      "types": Array [],
      "variants": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "loc": Object {
            "end": 32,
            "start": 19,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 24,
              "start": 19,
            },
            "value": "world",
          },
          "type": Object {
            "kind": "NamedType",
            "loc": Object {
              "end": 32,
              "start": 26,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 32,
                "start": 26,
              },
              "value": "String",
            },
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 34,
    "start": 0,
  },
}
`;

exports[`Schema Parser parses type with description string 1`] = `
Object {
  "block": false,
  "kind": "StringValue",
  "loc": Object {
    "end": 13,
    "start": 0,
  },
  "value": "Description",
}
`;
